File: task-export/src/main/scala/com/zxkw/userprofile/app/TaskExport.scala
package com.atguigu.userprofile.app

import java.util.Properties

import com.atguigu.userprofile.common.bean.TagInfo
import com.atguigu.userprofile.common.dao.TagInfoDAO
import com.atguigu.userprofile.common.util.{ClickhouseUtil, MyPropertiesUtil}
import org.apache.spark.SparkConf
import org.apache.spark.sql.{DataFrame, Row, SaveMode, SparkSession}

object TaskExport {
//  1  clickhouse 中建立宽表
//      1  一次性建 周期建？  周期性 2  程序中建
//  2  从hive中读取数据
//
//  3   写入到clickhouse中

  def main(args: Array[String]): Unit = {

    //  1  clickhouse 中建立宽表
    //      1  一次性建 周期建？  周期性 2  程序中建
     //  create table $tableName
     //  (uid String ,tagcode1 String ,tagcode2 String ...)
    //   engine = MergeTree
    //   order by uid

    val sparkConf: SparkConf = new SparkConf().setAppName("task_export_app")//.setMaster("local[*]")
    val sparkSession: SparkSession = SparkSession.builder().config(sparkConf).enableHiveSupport().getOrCreate()


    val taskId: String = args(0)
    val taskDate: String = args(1)

    val tableName=s"user_tag_merge_${taskDate.replace("-","")}"
    // 标签宽表的列 取决于 有多少个 启用状态的标签 --> 通过查询mysql中的tag_info+task_info
    val tagInfoList: List[TagInfo] = TagInfoDAO.getTagInfoListTaskOn()
    val tagCodeList: List[String] = tagInfoList.map(tagInfo=> s" ${tagInfo.tagCode.toLowerCase} String")
    val tagCodeSQL: String = tagCodeList.mkString(",")


    val properties: Properties = MyPropertiesUtil.load("config.properties")
    val userProfileDbName = properties.getProperty("user-profile.dbname")
    val wareHouseDbName = properties.getProperty("data-warehouse.dbname")
    val clickhouseUrl = properties.getProperty("clickhouse.url")

    var createTableSQL=
      s"""
         |     create table $tableName
         |       (  uid String ,$tagCodeSQL)
         |       engine = MergeTree
         |       order by uid
       """.stripMargin
    println(createTableSQL)
     ClickhouseUtil.executeSql(s"drop table if exists  $tableName")
     ClickhouseUtil.executeSql(createTableSQL)


    //  2  从hive中读取数据
   //rdd dataframe dataset
    println(s"select * from $tableName" )
    val dataFrame: DataFrame = sparkSession.sql(s"select * from $userProfileDbName.$tableName" )

    //driver  //executor

    //  3   写入到clickhouse中
   //   val rows: Array[Row] = dataFrame.collect()   // 从ex 提取到driver 中   无法分布式处理
   //  ClickhouseUtil.executeSql("insertxx" ,rows)
    println(s"写入clickhouse" )
    dataFrame.write.mode(SaveMode.Append)
        .option("batchsize",1000) //批量提交
        .option("isolationLevel", "NONE") //关闭事务
        .option("numPartitions",6)//设定并行度
       .option("driver","ru.yandex.clickhouse.ClickHouseDriver")
      .jdbc(clickhouseUrl,tableName,new Properties())


  }
}

File: task-export/src/main/java/TaskExport.java
import com.zxkw.userprofile.common.bean.TagInfo;
import com.zxkw.userprofile.common.constants.ConstCode;
import com.zxkw.userprofile.common.dao.TagInfoDao;
import com.zxkw.userprofile.common.util.ClickHouseUtil;
import com.zxkw.userprofile.common.util.MyPropsUtil;
import org.apache.spark.SparkConf;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.SparkSession;

import java.util.List;
import java.util.Properties;
import java.util.stream.Collectors;

/**
 * 为优化查询速度;成为即席查询,将合并好的宽表数据导出/迁移至clickhouse;如今估计流行doris
 * 1. 通过查询标签集合 , 获得字段名集合
 * 2. 表结构的平移,建一张clickhouse的表， 结构和hive的表字段一致
 * 3. 把数据写入clickhouse
 */
public class TaskExport {
    public static void main(String[] args) {

        SparkConf sparkConf = new SparkConf().setAppName("task_export_app").setMaster("local[*]");
        SparkSession sparkSession = SparkSession.builder().config(sparkConf).enableHiveSupport().getOrCreate();

        String taskId = args[0];
        String businessDate = args[1];

        String upDbName = MyPropsUtil.get(ConstCode.USER_PROFILE_DBNAME);
        String clickHouseUrl = MyPropsUtil.get(ConstCode.CLICKHOUSE_URL);

        //1. 通过查询标签集合 , 获得字段名集合
        List<TagInfo> tagInfoList = TagInfoDao.getTagInfoList();
        String fieldSQL =
                tagInfoList.stream().map(tagInfo -> tagInfo.getTagCode().toLowerCase() + " String").collect(Collectors.joining(","));

        //2. 表结构的平移,建一张clickhouse的表， 结构和hive的表字段一致
        // clickhouse的建表语句  表名  字段  引擎  主键  排序  分区  颗粒度8192
        // create table up_tag_merge_20200614
        //(
        //  uid String , 字段名 String ....
        // )
        // engine=MergeTree
        // order by uid
        String tableName = "up_tag_merge_"+businessDate.replace("-","");
        String dropTableSQL = " drop table if exists "+ upDbName+"."+tableName;
        System.out.println("dropTableSQL :" + dropTableSQL);

        ClickHouseUtil.executeSql(dropTableSQL);

        String createTableSQL = " create table if not exists " + upDbName + "." + tableName +
                " (" +
                " uid String , " + fieldSQL +
                " )" +
                " engine=MergeTree" +
                " order by uid";
        System.out.println("createTableSQL : "+ createTableSQL);
        ClickHouseUtil.executeSql(createTableSQL);

        //3. 把数据写入clickhouse
        // hiveTable  =>  dataFrame  =>  jdbc

        Dataset<Row> dataset = sparkSession.sql("select * from " + upDbName + "." + tableName);

        dataset.write().mode(SaveMode.Append)
                .option("driver", "ru.yandex.clickhouse.ClickHouseDriver")
                .option("batchsize" ,200)
                .option("isolationLevel", "NONE")
                .option("numPartitions",4)
                .jdbc(clickHouseUrl , tableName , new Properties());
    }
}
